### ISSUE TITLE
Doc printer is not handling `trim` correctly

### ISSUE THREAD

**Issue by fisker** [prettier/prettier#14152]
According to [documentation of `trim` command](https://github.com/prettier/prettier/blob/main/commands.md#trim), it should only trim "all the indentation on the current line. ".

So [`["a ", trim, 'b']`](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBtAOiAhgAkwGhxgCcBLAW0IHIAjKgXRHxAgAcZToBnZULY4hADuABX4IeKLABshWAJ49mNYljABrODADKWcnAAypKHGQAzGVzjLVGrdtZrjAc2QkArtZBxyNOABN-AIMsKGd3LGc4ADEIYnIsGA4w5Gx3GAgmEAALGHJpAHVs0nguRzA4bQkS0gA3EvlUsC4lEGMrYhgRVWcE80svACsuAA9tF2k4AEV3CHh+6StmR2IO1P8IMABaOBHWaTi4YizWMlgC0n8YbOQADgAGZcErAtVWVNO4DtrTZgBHWbwbpsSTYLhbEwBAJZYhwAGkWHdSJ9JAWRZeKzkUhuYieZhcCZwACCSTINHScBERyMJgWSxABLCkxmc1MqIGzBgWBoFyuNyQACZOapSNIXABhCDkFHeLgAViy7isABVuZI0fTap4AJJQIKwbRgMjsIl67QweSTOlwAC+NqAA) should print as `a b`, not `ab`.

Found this issue when trying to fix #6793. This also hiding bugs where `softline` should be used instead of `line`.

In the doc printer, we use an array to save strings, it's harder to location indention (not really, but if we want it prefect).
Maybe it's easier to use a string (not finished PR #13882).

**Comment by thorn0** [comment]
The next sentence in the documentation says "Should be placed after a line break." 🤓
If you place it after a line break, it trims the indentation. Otherwise, it does something else (in particular, it might be not handled correctly by `fits` if placed in the middle of a line).

**Comment by fisker** [comment]
Interesting. But it seems it's called automatically *BEFORE* line break.. https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/document/doc-printer.js#L537

**Comment by thorn0** [comment]
True, but the documentation is only about the `trim` doc, not about this function.
As for #6793, it might be better to invent a way to escape those two spaces, so that the `trim` function wouldn't touch them and work as before in the rest of the cases.

**Comment by fisker** [comment]
Looks like you [changed the documentation](https://github.com/prettier/prettier/commit/58173be8b8e716def89ea2594672abfc842b681b#diff-36a23b056c0b88915240fda511cce74477567f8b3891b96d0764f3232098c0fdR255), it doesn't mean to trim indentation,  there is a [test](https://github.com/prettier/prettier/pull/4772/files#diff-34cfe7771de5f0cd1526dfc800d035d5e49373069afe4ac9a92f5249376045e5R50) confirms.

**Comment by thorn0** [comment]
Indeed it was me. I did this because `fits` doesn't support [some scenarios](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBzAThArgBwBQDaAOlAASlEh6UA0J5A5Ocy68w3WaTOgJYC2nRgDMIEDvVIAbXlDhCKIAJS0SAXSUgaICDhi9oAZ2SgAhukwB3AArmExlKamXTAT2PaARulNgA1nAwAMqm-HAAMrJwyMJOhvIg3r4BwTi+sqjIPFgJcPyecAAmhUXhplCoWKaocABiEOj8pjD6FcggplgwEFogABYw-FIA6n288IZpYHBB9uO8AG7jru1ghh4gsvHoMNY+qE0xcQkAVoYAHkEZUnAAilgQ8EdS8dpp6NvthRBgALRw5xwUgacHQvRwfFgw14hRgfWQAE43ph4sMfDh2hC4NsFtFtABHB7wPa6BwdQy-ORFIq9dBwQm8Ol7aqHJCxF4JeL8XhZdA5bSGa5wACCLT4ni6cGsoMicmerxAgoqN3uj2ibOO2hgpk80Nh8KQACYtT5eDIKgBhCD8VkgbEAVl6WHiABUdQ52QqFjkAJJQEqwIJgPh6YX+oIwVw3eVwAC+saAA) where `trim` is in the middle of the line.

**Comment by thorn0** [comment]
> it doesn't mean to trim indentation

Of course, it does. It was added for formatting preprocessor instructions.

**Comment by fisker** [comment]
Should we change this line to only trim indentation?

https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/document/doc-printer.js#L537

**Comment by thorn0** [comment]
Probably. That'd be a kind of breaking change, something for v3. Prettier trims trailing whitespace this way, and we don't know who and how relies on this behavior. Seems like we would have to reimplement this trimming in multiple places even in Prettier's own printers, but probably it's not a bad thing. Overall, I'm nor sure.

**Comment by thorn0** [comment]
BTW, I tried to reimplement printing `trim` and new lines once. The idea was to defer printing (flushing) the indentation until it's needed. I don't remember why I abandoned that attempt. But if we go this way https://github.com/prettier/prettier/issues/14152#issuecomment-1377689665, then that idea can be used (at least it seems so) to get rid of the `trim` function completely.
